kpatch-build/create-diff-object.c

```diff
@@ -623,153 +680,86 @@ static bool kpatch_line_macro_change_only_x86_64(struct section *sec)
-		insn_get_opcode(&insn1);
-		insn_get_opcode(&insn2);
-		if (!(insn1.opcode.value == 0xba && insn2.opcode.value == 0xba) &&
-		    !(insn1.opcode.value == 0xbe && insn2.opcode.value == 0xbe))
-			return false;
-
 		/*
-		 * Verify zero or more string relas followed by a
-		 * warn_slowpath_* or another similar rela.
+		 * Here we found a difference between two instructions of the
+		 * same length.  Only ignore the change if:
+		 *
+		 *   a) the instructions match a known pattern of a '__LINE__'
+		 *      macro immediate value which was embedded in the
+		 *      instruction; and
+		 *
+		 *   b) the instructions are followed by certain expected
+		 *      relocations.
 		 */
-		found = 0;
-		list_for_each_entry(rela, &sec->rela->relas, list) {
-			if (rela->offset < offset + length)
-				continue;
-			if (rela->string)
-				continue;
-			if (!strncmp(rela->sym->name, "__warned.", 9) ||
-			    !strncmp(rela->sym->name, "__already_done.", 15))
-				continue;
-			if (!strncmp(rela->sym->name, "warn_slowpath_", 14) ||
-			   (!strcmp(rela->sym->name, "__warn_printk")) ||
-			   (!strcmp(rela->sym->name, "__might_sleep")) ||
-			   (!strcmp(rela->sym->name, "___might_sleep")) ||
-			   (!strcmp(rela->sym->name, "__might_fault")) ||
-			   (!strcmp(rela->sym->name, "printk")) ||
-			   (!strcmp(rela->sym->name, "lockdep_rcu_suspicious"))) {
-				found = 1;
-				break;
-			}
-			return false;
-		}
-		if (!found)
-			return false;
-
-		lineonly = 1;
-	}
-
-	if (!lineonly)
-		ERROR("no instruction changes detected for changed section %s",
-		      sec->name);
-
-	return true;
-}
-
-#define PPC_INSTR_LEN 4
-#define PPC_RA_OFFSET 16
-
-static bool kpatch_line_macro_change_only_ppc64le(struct section *sec)
-{
-	unsigned long start1, start2, size, offset;
-	unsigned int instr1, instr2;
-	struct rela *rela;
-	int lineonly = 0, found;
-
-	if (sec->status != CHANGED ||
-	    is_rela_section(sec) ||
-	    !is_text_section(sec) ||
-	    sec->sh.sh_size != sec->twin->sh.sh_size ||
-	    !sec->rela ||
-	    sec->rela->status != SAME)
-		return false;
 
-	start1 = (unsigned long)sec->twin->data->d_buf;
-	start2 = (unsigned long)sec->data->d_buf;
-	size = sec->sh.sh_size;
-	for (offset = 0; offset < size; offset += PPC_INSTR_LEN) {
-		if (!memcmp((void *)start1 + offset, (void *)start2 + offset,
-			    PPC_INSTR_LEN))
-			continue;
+		if (!insn_is_load_immediate(kelf, insn1) ||
+		    !insn_is_load_immediate(kelf, insn2))
+			return false;
 
-		instr1 = *(unsigned int *)(start1 + offset) >> PPC_RA_OFFSET;
-		instr2 = *(unsigned int *)(start2 + offset) >> PPC_RA_OFFSET;
+		found = false;
+		list_for_each_entry(r, &sec->rela->relas, list) {
 
-		/* verify it's a load immediate to r5 */
-		if (!(instr1 == 0x38a0 && instr2 == 0x38a0))
-			return false;
+			if (r->offset < offset + insn1_len)
+				continue;
 
-		found = 0;
-		list_for_each_entry(rela, &sec->rela->relas, list) {
-			if (rela->offset < offset + PPC_INSTR_LEN)
+			rela = toc_rela(r);
```

Comment:
Ok, **toc_rela()** should just return the **rela** back for non-ppc64le.

Reply:
Right, and also for the ppc64le case where the rela isn't a toc-style nested rela.